### PR TITLE
Ubuntu install directions

### PR DESCRIPTION
**Pull request by jeffsikes**
Condensed instructions from my blog post on how to install FediBuzz on an Ubuntu machine, as requested by astro. 

Updated to address concerns from @SuperSandro2000 to exclude details not directly relevant to the install, uniform references for the term PostgreSQL, and clarify sections related to SSL Certificates and the PostgreSQL installation and configuration.

For some of the smaller changes, I went ahead and marked them as resolved. I wasn't sure if that was appropriate etiquette in this repository, so left the rest as open comments for your review.

Thanks!

### CODE REVIEW

File: INSTALL-UBUNTU.md [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+The following ports need to be open on the server running FediRelay. 
+
+```
+## Default is 3000 in the FediBuzz docs, change as needed
+sudo ufw allow 3000
+## Allow SSH traffic so you can connect - consider limiting to specific IPs
+sudo ufw allow 22
+
+## When ready...
+sudo ufw enable
```
Comment (SuperSandro2000) [inline INSTALL-UBUNTU.md:48]: Ubuntu users likely use ufw but enabling it here is a bit out of scope of this tutorial

File: INSTALL-UBUNTU.md [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+
+If you promote this alternative route and many individuals start connecting to your relay, it will cause more outgoing traffic and queue processing, therefore increasing your hardware requirements.
+
+# Domain Name
+As with most fediverse projects, you're going to need a domain. In this particular instance, I used a subdomain of https://relay.{yourdomain}.
+
+# NGINX Proxy Manager Config 
+You'll need an SSL certificate for at least the sub-domain. NGINX Proxy Manager has this built in, but you may need some additional assistance depending on your configuration.
+
+I needed a wildcard SSL for my domain, and used the PorkBun API to make that work. Here's a [blog post about Cloudflare](https://blog.jverkamp.com/2023/03/27/wildcard-lets-encrypt-certificates-with-nginx-proxy-manager-and-cloudflare/), but the setup is very similar for PorkBun.
```
Comment (SuperSandro2000) [inline INSTALL-UBUNTU.md:34]: Totally unrelated

File: INSTALL-UBUNTU.md [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+
+Websockets may not be required, but I enabled it in NGINX Proxy Manager during configuration.
```
Comment (SuperSandro2000) [inline INSTALL-UBUNTU.md:36]: Please clarify

File: INSTALL-UBUNTU.md [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+
+One caveat here. FediBuzz has an option for individual users to utilize relays as well by "following" a relay actor, like `@tag-dogsofmastodon@relay.com`. 
+
+If you promote this alternative route and many individuals start connecting to your relay, it will cause more outgoing traffic and queue processing, therefore increasing your hardware requirements.
+
+# Domain Name
+As with most fediverse projects, you're going to need a domain. In this particular instance, I used a subdomain of https://relay.{yourdomain}.
+
+# NGINX Proxy Manager Config 
+You'll need an SSL certificate for at least the sub-domain. NGINX Proxy Manager has this built in, but you may need some additional assistance depending on your configuration.
```
Comment (SuperSandro2000) [inline INSTALL-UBUNTU.md:32]: This should be done in a generic way. People should be instructed to add this to their reverse proxy with a working tls config and they should consult the software doc or other tutorials how to do this.

File: INSTALL-UBUNTU.md [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+The official fedi.buzz relay traffics around 330 public instance streams for around 3,500 unique followers (some may have multiple relay requests) as of April 2024 with a similar configuration.
+
+If you're connecting to the fedi.buzz relay and perhaps one or two others on your own relay, this should be more than enough.
+
+One caveat here. FediBuzz has an option for individual users to utilize relays as well by "following" a relay actor, like `@tag-dogsofmastodon@relay.com`. 
+
+If you promote this alternative route and many individuals start connecting to your relay, it will cause more outgoing traffic and queue processing, therefore increasing your hardware requirements.
+
+# Domain Name
+As with most fediverse projects, you're going to need a domain. In this particular instance, I used a subdomain of https://relay.{yourdomain}.
```
Comment (SuperSandro2000) [inline INSTALL-UBUNTU.md:29]: The first sentence is kinda wrong. You don't need a new domain, only a subdomain and that's true for almost any selfhosted project unless you are using subpaths which are usually finicky

File: INSTALL-UBUNTU.md [hunk 6]
```diff
@@ -0,0 +1,257 @@
+# Objective
+Install the FediBuzz relay using a virtual machine running Ubuntu Server 22.04.4 LTS. If you are familiar with NixOS/Flakes, then [installing the NixOS module](https://github.com/astro/buzzrelay?tab=readme-ov-file#build) for this is by far the easier route! 
+
+But for those of us unfamiliar with NixOS, here's another option. 
+
+# Server Configuration
+My system consisted of the following:
+
+* Router Configured to forward 80, 443 through a firewall to NGINX Proxy Manager
```
Comment (SuperSandro2000) [inline INSTALL-UBUNTU.md:9]: That is all nice but not really relevant for a guide here. We don't want to project to people that they need proxmox or ngonx proxy manager

File: INSTALL-UBUNTU.md [hunk 7]
```diff
@@ -0,0 +1,257 @@
+# Objective
+Install the FediBuzz relay using a virtual machine running Ubuntu Server 22.04.4 LTS. If you are familiar with NixOS/Flakes, then [installing the NixOS module](https://github.com/astro/buzzrelay?tab=readme-ov-file#build) for this is by far the easier route!
```
Comment (SuperSandro2000) [inline INSTALL-UBUNTU.md:2]: Mentioning the exact Ubuntu release will just make this guide look dated in no time. The steps should be generic enough that they very likely continue to work in the future.

File: INSTALL-UBUNTU.md (hunk 5 above)
Comment (jeffsikes) [reply INSTALL-UBUNTU.md]: Reworded to be clearer you can use a sub-domain.

File: INSTALL-UBUNTU.md (hunk 3 above)
Comment (jeffsikes) [reply INSTALL-UBUNTU.md]: I wasn't sure if enabling websockets was required. Verified it is not, so removed that from the doc.

File: INSTALL-UBUNTU.md (hunk 1 above)
Comment (jeffsikes) [reply INSTALL-UBUNTU.md]: Section removed.

File: INSTALL-UBUNTU.md (hunk 2 above)
Comment (jeffsikes) [reply INSTALL-UBUNTU.md]: These were part of my personal notes, as stated in the pull request. It was pertinent to me as one of the steps required to get this project working. 

I've removed this, agreed that it should stay in my personal journal entry of the process.

File: INSTALL-UBUNTU.md [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+sudo ufw allow 22
+
+## When ready...
+sudo ufw enable
+```
+
+# Initial Server Installs
+These packages are required for rust / cargo to work.
+```
+sudo apt-get update
```
[the remark this answers is not in the document]
Comment (jeffsikes) [reply INSTALL-UBUNTU.md:29]: Updated to apt-get and consolidated

File: INSTALL-UBUNTU.md [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+sudo apt install pkg-config
```
[the remark this answers is not in the document]
Comment (jeffsikes) [reply INSTALL-UBUNTU.md]: Consolidated as requested.

File: INSTALL-UBUNTU.md [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+sudo apt install curl
+```
+
+curl was already installed for me.
+
+### Rust and related tooling install
+Rust is already installed on Ubuntu, but not compatible with rustup. Remove it.
+
+```
+sudo apt remove rustc cargo
```
[the remark this answers is not in the document]
Comment (jeffsikes) [reply INSTALL-UBUNTU.md]: Removed and replaced wtih a link to the official rust installation docs.

File: INSTALL-UBUNTU.md [hunk 11]
```diff
@@ -0,0 +1,257 @@
...
+source $HOME/.cargo/env
+```
+
+# Pull GitHub Repo 
+```
+git clone https://github.com/astro/buzzrelay.git
+```
+
+# Postgres SQL
+Install a PostGresSQL database, I followed this guide.
```
[the remark this answers is not in the document]
Comment (jeffsikes) [reply INSTALL-UBUNTU.md]: All references are now shown as "PostgreSQL" throughout the document. This matches the terminology used in the README.md

File: INSTALL-UBUNTU.md [hunk 12]
```diff
@@ -0,0 +1,257 @@
...
+```
+createuser --interactive
+```
+
+Role (user) to add: relay as superuser
+Password wasn't set, so from a postgres prompt:
+
+`ALTER USER relay WITH PASSWORD 'your-secure-password';`
+
+Then create database:
```
[the remark this answers is not in the document]
Comment (jeffsikes) [reply INSTALL-UBUNTU.md]: The PostgreSQL section has been updated to simplify the process as requested. Initially, the user was created with a superuser role, I removed that. This requires providing the GRANT statements on the buzzrelay database for the relay user.

File: INSTALL-UBUNTU.md [hunk 13]
```diff
@@ -0,0 +1,257 @@
...
+ALTER DEFAULT PRIVILEGES IN SCHEMA public GRANT ALL PRIVILEGES ON SEQUENCES TO relay; 
+ALTER DEFAULT PRIVILEGES IN SCHEMA public GRANT ALL PRIVILEGES ON FUNCTIONS TO relay;
+GRANT USAGE, CREATE ON SCHEMA public TO relay;
+```
+
+## Querying the database
+A cheat sheet for getting to the database.
+
+```
+sudo -i -u postgres
```
[the remark this answers is not in the document]
Comment (jeffsikes) [reply INSTALL-UBUNTU.md]: These lines have been modified to show how to connect to the database as the user that was created, which validates that the user has the appropriate permissions for the application.

File: INSTALL-UBUNTU.md [hunk 14]
```diff
@@ -0,0 +1,257 @@
...
+Set it to your domain. I used the sub-domain format of "relay.{yourdomain}"
+### Private Key File
+Generate a new RSA key pair for signing ActivityPub messages. Note using this command also sets the appropriate permissions values.
+
+```bash
+openssl genrsa -out private-key.pem 4096 
+openssl rsa -in private-key.pem -pubout -out public-key.pem
+```
+
+## PostGres Password
```
[the remark this answers is not in the document]
Comment (jeffsikes) [reply INSTALL-UBUNTU.md]: All references are now shown as "PostgreSQL" throughout the document. This matches the terminology used in the README.md

File: INSTALL-UBUNTU.md [hunk 15]
```diff
@@ -0,0 +1,257 @@
...
+Edit the new service file, change the working directory to your own location.
+
+```nano
+[Unit]
+Description=Buzzrelay Rust Application
+After=network.target
+
+[Service]
+Type=simple
+User=box464
```
[the remark this answers is not in the document]
Comment (jeffsikes) [reply INSTALL-UBUNTU.md]: I removed the directions for configuring automated startup completely and will leave the preferred options for startup to the user. If you would prefer to include something here and have a suggestion for a replacement, happy to include it.

File: INSTALL-UBUNTU.md [hunk 16]
```diff
@@ -0,0 +1,257 @@
...
+WorkingDirectory=/home/box464/buzzrelay
```
[the remark this answers is not in the document]
Comment (jeffsikes) [reply INSTALL-UBUNTU.md]: I removed the directions for configuring automated startup completely and will leave the preferred options for startup to the user. If you would prefer to include something here and have a suggestion for a replacement, happy to include it.

File: INSTALL-UBUNTU.md (hunk 4 above)
Comment (jeffsikes) [reply INSTALL-UBUNTU.md]: This section was renamed SSL Certificate and revamped. I've made the instructions more generic as requested.

**Comment by astro** [comment]
Thanks!

**Pull request merged**